### DISCUSSION TITLE
"public" vs "private" configuration values

### DISCUSSION THREAD

**Discussion by szabgab** [rwf2/Rocket#2842] [Questions]
In the application I am writing I need to store both private API keys of external sites (E.g. sendgrid to send out emails) and public information as well (eg. the Google analytics code I need to include in every page.) I'd need to pass the public values to `Template::render` but I think it would be better not to pass the private values to avoid accidentally publishing them.

I could include pass the individual public values to each call of `Template::render` but as the number of these public configuration values might grow it would not be nice to pass the same 2-3.. parameters to every function.

How would you solve this?

**Comment by the10thWiz** [comment]
There are two primary options. If the public information isn't dependent on the content of the page, you can inject it as a custom function directly into the templating engine, otherwise you could wrap it into a struct (as long as it implements Serialize, you can pass it as a single parameter to context!).

**Reply by szabgab** [reply]
Thanks.

Could you give me some more hints how to "inject it as a custom function directly in the templating engine"?

**Reply by the10thWiz** [reply]
See https://api.rocket.rs/v0.5/rocket_dyn_templates/struct.Template#method.custom.
